pkg/engine/variables/variables_test.go

```diff
@@ -0,0 +1,278 @@
+package variables
+
+import (
+	"encoding/json"
+	"reflect"
+	"testing"
+
+	"github.com/nirmata/kyverno/pkg/engine/context"
+)
+
+func Test_variableSubstitutionValue(t *testing.T) {
+
+	resourceRaw := []byte(`
+	{
+		"metadata": {
+			"name": "temp",
+			"namespace": "n1"
+		},
+		"spec": {
+			"namespace": "n1",
+			"name": "temp1"
+		}
+	}
+		`)
+	patternMap := []byte(`
+	{
+		"spec": {
+			"name": "{{request.object.metadata.name}}"
+		}
+	}
+	`)
+
+	resultMap := []byte(`{"spec":{"name":"temp"}}`)
+
+	var pattern, resource interface{}
+	json.Unmarshal(patternMap, &pattern)
+	json.Unmarshal(resourceRaw, &resource)
+
+	// context
+	ctx := context.NewContext()
+	ctx.AddResource(resourceRaw)
+	value := SubstituteVariables(ctx, pattern)
+	resultRaw, err := json.Marshal(value)
+	if err != nil {
+		t.Error(err)
+	}
+	if !reflect.DeepEqual(resultMap, resultRaw) {
+		t.Error("result does not match")
+	}
+}
+
+func Test_variableSubstitutionValueOperatorNotEqual(t *testing.T) {
+
+	resourceRaw := []byte(`
+	{
+		"metadata": {
+			"name": "temp",
+			"namespace": "n1"
+		},
+		"spec": {
+			"namespace": "n1",
+			"name": "temp1"
+		}
+	}
+		`)
+	patternMap := []byte(`
+	{
+		"spec": {
+			"name": "!{{request.object.metadata.name}}"
+		}
+	}
+	`)
+	resultMap := []byte(`{"spec":{"name":"!temp"}}`)
+
+	var pattern, resource interface{}
+	json.Unmarshal(patternMap, &pattern)
+	json.Unmarshal(resourceRaw, &resource)
+
+	// context
+	ctx := context.NewContext()
+	ctx.AddResource(resourceRaw)
+	value := SubstituteVariables(ctx, pattern)
+	resultRaw, err := json.Marshal(value)
+	if err != nil {
+		t.Error(err)
+	}
+	t.Log(string(resultRaw))
+	t.Log(string(resultMap))
+	if !reflect.DeepEqual(resultMap, resultRaw) {
+		t.Error("result does not match")
+	}
+}
+
+func Test_variableSubstitutionValueFail(t *testing.T) {
+
+	resourceRaw := []byte(`
+	{
+		"metadata": {
+			"name": "temp",
+			"namespace": "n1"
+		},
+		"spec": {
+			"namespace": "n1",
+			"name": "temp1"
+		}
+	}
+		`)
+	patternMap := []byte(`
+	{
+		"spec": {
+			"name": "{{request.object.metadata.name1}}"
+		}
+	}
+	`)
+	resultMap := []byte(`{"spec":{"name":null}}`)
```

Comment:
Is this `null` an empty string or a field to ignore? If the referred path does not has a value, should we drop this key in the result, To have something like `resultMap := []byte(``)`?

Take the annotation as an example: `a: null`, it would be converted to nil and fails k8s validation check:
````
error validating data: unknown object type "nil" in Pod.metadata.annotations.a; if you choose to ignore these errors, turn validation off with --validate=false
````

Reply:
@realshuting 
This is difficult from the variable substitution point of view, as the return value can be a scalar, object or evaluate or incorrect path that results in nil. We don't check the expected type, as the variable can refer to object type but the value expected is of type string. This will be an invalid substitution and result in an error.

Regarding mutation, when we create the mutation patches we parse over the structure and build the mutation patch. This is when we can handle the above scenario. This depends on default or nil values for each data type, for strings nil value is empty string while for a pointer it will be nil and an object the empty object.

```
overlay:
  v1:
    var: nil
```
The above should be handled while generation mutation patches. I'll create a child-issue to this for handling the scenarios in mutation to you.